### DISCUSSION TITLE
"next dev --turbo" fails in WASM with error (`turbo.createProject` is not supported by the wasm bindings)

### DISCUSSION THREAD

**Discussion by theshadowagent** [vercel/next.js#75665] [Ideas]
### Link to the code that reproduces this issue

https://stackblitz.com/edit/stackblitz-starters-yew3c9

### To Reproduce

1. Start `next` dev server with Turbopack (`next dev --turbo`) in WASM (e.g. using webcontainers.io)
2. The `dev` command fails with error
```
~/projects/stackblitz-starters-yew3c9
❯ npm install

added 10 packages, and changed 6 packages in 34s

144 packages are looking for funding
  run `npm fund` for details

~/projects/stackblitz-starters-yew3c9 34s
❯ npx next dev --turbo
  ▲ Next.js 15.0.0-canary.171 (turbo)
  - Local:        http://localhost:3000

 ✓ Starting...
   Downloading swc package @next/swc-wasm-nodejs... to /home/.cache/next-swc
Error: `turbo.createProject` is not supported by the wasm bindings.
    at Object.createProject (/home/projects/stackblitz-starters-yew3c9/node_modules/next/dist/build/swc/index.js:808:31)
    at createHotReloaderTurbopack (/home/projects/stackblitz-starters-yew3c9/node_modules/next/dist/server/dev/hot-reloader-turbopack.js:121:42)
    at async startWatcher (/home/projects/stackblitz-starters-yew3c9/node_modules/next/dist/server/lib/router-utils/setup-dev-bundler.js:164:38)
    at async setupDevBundler (/home/projects/stackblitz-starters-yew3c9/node_modules/next/dist/server/lib/router-utils/setup-dev-bundler.js:814:20)
    at async Span.traceAsyncFn (/home/projects/stackblitz-starters-yew3c9/node_modules/next/dist/trace/trace.js:157:20)
    at async initialize (/home/projects/stackblitz-starters-yew3c9/node_modules/next/dist/server/lib/router-server.js:87:30)
    at async Server.eval (/home/projects/stackblitz-starters-yew3c9/node_modules/next/dist/server/lib/start-server.js:266:36)
```

### Current vs. Expected behavior

I expected `next dev --turbo` to work in WASM mode so I could have fast hot reloading, but it throws this error

### Provide environment information

```bash
❯ npx next info

Operating System:
  Platform: linux
  Arch: x64
  Version: Ubuntu 20.04.0 LTS
  Available memory (MB): NaN
  Available CPU cores: 8
Binaries:
  Node: 18.20.3
  npm: 10.2.3
  Yarn: 1.22.19
  pnpm: 8.15.6
Relevant Packages:
  next: 15.0.0-canary.171 // Latest available version is detected (15.0.0-canary.171).
  eslint-config-next: 13.5.1
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Turbopack

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

I tested my reproduction against "15.0.0-canary.171" (latest canary release), and "14.2.13" (latest stable release). On stable it outputs a different error: `TypeError: bindings.turbo.createProject is not a function`.

I found the code that throws this: https://github.com/vercel/next.js/blob/3ed9f4b3f4d1f8b431ff04d4e6a45a949680a31f/packages/next/src/build/swc/index.ts#L1063

Is wasm bindings support planned for turbopack anytime soon? Was very bummed to see this, since it considerably slows down next.js in browser IDEs.

**Comment by samcx** [comment]
@theshadowagent Thank you for submitting an issue!

I was not able to <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> this.

![CleanShot 2024-09-30 at 23 28 47@2x](https://github.com/user-attachments/assets/1eb3ff98-2d53-4e99-87d6-9e6e1706de5e)

**Comment by abhi12299** [comment]
the link provided works fine - no WASM errors on stackblitz

**Comment by theshadowagent** [comment]
@samcx @abhi12299 
Here's an update on how to reproduce:
1. Open the Stackblitz link I provided & stop the existing terminal command using Ctrl+C. The issue is that Stackblitz runs `next dev` without `--turbo` flag by default (specifically, it runs `npm install && npx next dev`)
2. Manually run `npm install && npm run dev` which runs the `next dev --turbo` command and fails
<img width="1511" alt="image" src="https://github.com/user-attachments/assets/160114f1-15ae-480d-9048-cec3f4aa3d61">

Apologies for confusing you here – I couldn't find a way to change the default run command in Stackblitz

Thanks for the swift response!

**Comment by theshadowagent** [comment]
Also wanted to add about why it's important for me:
My default next14 + shadcn + tailwind project takes ~60 seconds to boot in Webcontainers without `--turbo` flag, and 83% of the time is spent on the compile step, which I suspect takes so long due to using Webpack instead of Turbopack.

**Comment by timneutkens** [comment]
Turbopack does not have a WebAssembly build currently so I'm not expecting it to work on Stackblitz right now.

It's not unexpected that running in Stackblitz is slower than locally because it will have to download the WebAssembly binary and then use that instead of using the native bindings.

**Comment by theshadowagent** [comment]
@timneutkens thanks, I understand. Is WebAssembly build for Turbopack on the roadmap somewhere, or should I switch to server environment?

On WebAssembly binary: I've managed to reduce the time by 2x via adding `@next/swc-wasm-nodejs` to my `package.json`, but that only worked with `next@13.5.1`. `next@14.2.14` still downloads the binary (at least according to console) even if it's present in package.json. 

Still, compile step takes about 10s, so WebAssembly build for Turbopack would be amazing.
See [bolt.new](https://bolt.new) for why it might be important

**Comment by timneutkens** [comment]
We're currently focused on making sure that as many people as possible get the benefits of Turbopack, which means prioritizing development / builds with the native binding, after Turbopack development and build are stable we can investigate WebAssembly builds, though it's not straightforward to add this because of the premise of the compiler, which includes file watching, disk access (read/write) etc.

**Comment by theshadowagent** [comment]
Got it, makes total sense! I think I'll switch to server VMs for now. I'm curious if this release helps to progress on this: https://deno.com/blog/rusty-v8-stabilized

**Comment by illegitimate-egg** [comment]
Bumping this, except for me rather than not working on stackblitz it doesn't work on my FreeBSD laptop

**Comment by ahmedkhabar** [comment]
Finally, I fix it - the solution was so simple: just upgrade the Node.js version using [NVM](https://github.com/nvm-sh/nvm) with this command `nvm install NEW_VERSION --reinstall-packages-from=OLD_VERSION` and then I switch to the new Node version using `nvm use 23.4.0` and change the version number "23.4.0" with the installed version.

Edit:
use lts instead because the latest version of Node js have some issues with prisma.
```nvm install lts```

**Comment by TonyRice** [comment]
I'm currently running the next boilerplate on a termux install. I'm running the latest version of nodejs built for arm64

I get this same error on a fresh install for the boilerplate project.

**Comment by TonyRice** [comment]
You can optionally disable turbo and use webpack.

**Comment by 1sh4sk** [comment]
i ran into this error too i guess it happened because while creating the nextjs app some packages are not installed correctly what i did is delete the project and create it from the start and the error no longer exists

solution:
delete the nextjs project you just created
and create a new one from the start by using the command npx create-next-app@latest

**Comment by Badbird5907** [comment]
<img src="https://github.com/user-attachments/assets/ce282e6c-c93d-4b79-89b7-e57a97e9f775" width="400" />

Hi, I am getting this error with create-next-app@latest and next@latest (15.1.6).
I'm using Stackblitz's [webcontainer](https://webcontainers.io/).


Edit: can [reproduce](https://stackblitz.com/edit/stackblitz-starters-meyodqb1?file=package.json) in stackblitz

**Comment by mhr1380** [comment]
I had the same issue, and this worked for me: 🚀
1️⃣ Delete the .next and node_modules folders 🗑️
2️⃣ Run npm install (or yarn install) 📦
3️⃣ Restart the development server 🔄

Hope this helps! ✨

**Reply by JonyChann** [reply]
This worked for me.
I pushed my repo from a PC(x64) and cloned into a Mac(arm64), alone with .next and node_modules.
Guess npm didn't recognize the packages from another architecture.
Thanks @mhr1380

**Comment by Sani071** [comment]
> I had the same issue, and this worked for me: 🚀 1️⃣ Delete the .next and node_modules folders 🗑️ 2️⃣ Run npm install (or yarn install) 📦 3️⃣ Restart the development server 🔄
> 
> Hope this helps! ✨

It worked for me. 
Thanks @mhr1380

**Comment by goleary** [comment]
@samcx Just curious why this issue was converted to a discussion?

This issue remains today with the latest version of next (15.5.2).

Repro here: https://stackblitz.com/edit/stackblitz-starters-tjdzkbpu

- [open](https://stackblitz.com/edit/stackblitz-starters-tjdzkbpu)
- cancel the running terminal command
- run npm run dev

```
[Error: `turbo.createProject` is not supported by the wasm bindings.]

**Comment by p-efimenko** [comment]
To resolve this issue, just try to install **Rust** on Windows

**Comment by parsamoloody** [comment]
I get an error when starting the app npm run dev|

 ```
✓ Starting...
   Skipping creating a lockfile at C:\projects\git\DigitalMenu2\app\.next\dev\lock because we're using WASM bindings
Error: `turbo.createProject` is not supported by the wasm bindings.
    at ignore-listed frames
```

**Comment by parsamoloody** [comment]
Make sure your installed nodejs version is correct and does not have any conflicts with your device.

**Comment by elbasel-404** [comment]
Try adding the "--webpack" flag
